pkg/engine/internal/proto/wirepb/wirepb.proto

```diff
@@ -0,0 +1,198 @@
+// wirepb.proto holds types for the wire protocol used for communication
+// between scheduler and workers in the query engine.
+syntax = "proto3";
+
+package loki.wire;
+
+import "gogoproto/gogo.proto";
+import "google/protobuf/duration.proto";
+import "google/protobuf/timestamp.proto";
+import "pkg/engine/internal/proto/expressionpb/expressionpb.proto";
+import "pkg/engine/internal/proto/physicalpb/physicalpb.proto";
+import "pkg/engine/internal/proto/ulid/ulid.proto";
+
+option go_package = "github.com/grafana/loki/v3/pkg/engine/internal/proto/wirepb";
+
+message Frame {
+  oneof kind {
+    AckFrame ack = 1;
+    NackFrame nack = 2;
+    DiscardFrame discard = 3;
+    MessageFrame message = 4;
+  }
+}
+
+message AckFrame {
+  uint64 id = 1;
+}
+
+message NackFrame {
+  uint64 id = 1;
+  string error = 2;
+}
+
+message DiscardFrame {
+  uint64 id = 1;
+}
+
+message MessageFrame {
+  uint64 id = 1;
+
+  oneof kind {
+    WorkerReadyMessage worker_ready = 2;
+
+    TaskAssignMessage task_assign = 3;
+    TaskCancelMessage task_cancel = 4;
+    TaskFlagMessage task_flag = 5;
+    TaskStatusMessage task_status = 6;
+
+    StreamBindMessage stream_bind = 7;
+    StreamDataMessage stream_data = 8;
+    StreamStatusMessage stream_status = 9;
+  }
+}
+
+message WorkerReadyMessage {}
+
+// TaskAssignMessage is sent by the scheduler to a worker when there is a
+// task to run.
+message TaskAssignMessage {
+  Task task = 1;
+
+  // StreamStates holds the most recent state of each stream that the task
+  // reads from. The key is the stream ULID.
+  map<string, StreamState> stream_states = 2;
```

Comment:
We can't use a ULID here? I don't think protobuf has any restrictions on map key types.

Reply:
Thanks, not sure how I missed that 🤦